pkg/policy/distillery/policy.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package distillery
+
+import (
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/policy"
+)
+
+type cachedSelectorPolicy struct {
+	users  map[Endpoint]struct{}
+	policy unsafe.Pointer
+}
+
+func newCachedSelectorPolicy() *cachedSelectorPolicy {
+	cip := &cachedSelectorPolicy{
+		users: make(map[Endpoint]struct{}),
+	}
+	cip.setPolicyLocked(policy.NewSelectorPolicy())
+	return cip
+}
+
+// getPolicy returns a reference to the SelectorPolicy that is cached.
+//
+// Users should treat the result as immutable state that MUST NOT be modified.
+func (cip *cachedSelectorPolicy) getPolicy() *policy.SelectorPolicy {
+	return (*policy.SelectorPolicy)(atomic.LoadPointer(&cip.policy))
+}
+
+// setPolicyLocked updates the reference to the SelectorPolicy that is cached.
+func (cip *cachedSelectorPolicy) setPolicyLocked(policy *policy.SelectorPolicy) {
+	atomic.StorePointer(&cip.policy, unsafe.Pointer(policy))
```

Comment:
if you are using atomic.Store/Load you don't need to hold any lock. This is a little bit confusing as well because `cip.policy` could simply be a `*policy.SelectorPolicy` which would be wrapped around a mutex that would save concurrent access of `users` and `policy` fields

Reply:
The lock isn't for the pointer storage, it's for keeping the `cip.policy` and `cip.revision` in sync. This might not be apparent until the later patches in the series.

The design of this (which I recognize is not documented, will fix) hands the `cachedSelectorPolicy` reference to an endpoint, which I want to be able to fetch the latest version of the policy without holding any locks (See the `Consume()` function). This is similar to to the approach used in #7539 , effectively "read and concurrent update"; the atomic provides a "publish" mechanism of an immutable copy of the most recently computed selector policy, and the policy calculation side can calculate and update the pointer without locking readers out.

Reply:
I've added additional comments to try to highlight this.

Reply:
And now I've implemented a change to streamline this.